### PR TITLE
Add BurstingRateLimiter for simulating super-bursty traffic.  

### PR DESCRIPTION
**Pull request by oschaaf**
Introduces a new `--burst-size` option. Implemented via `BurstingRateLimiter`, which accumulates acquisitions by the underlying configured `RateLimiter`, and starts releasing when the burst-size threshold is met. For example, `--burst-size 100 --rps 1000` will make the client release 100 requests every 100 milliseconds. 

Signed-off-by: Otto van der Schaaf <oschaaf@we-amp.com>

### CODE REVIEW

**Comment by oschaaf** [comment]
Rebased to master and squashed. Has an extra commit to fix the build that broke after the rebase:
- move to HeaderString::getStringView()

Force pushed the squashed version, but I'm leaving a note here to make sure I don't forget about the earlier ASAN failure in CI: 
https://circleci.com/gh/envoyproxy/nighthawk/209. This seems worth looking into, but it seems unlikely that it was related to this PR.

May need tweaking the coverage threshold in here, it may err because it didn't consider the test origin when I adjusted it earlier.

/assign @htuch

**Comment by oschaaf** [comment]
It looks like the hack to avoid runtime conflicts in the integration test for the client came back to haunt us :(
That test now flakes with `assert failure: thread_advancing_time_->isCurrentThreadId(). Details: time should only be advanced on one thread in the context of a test`.

**Comment by oschaaf** [comment]
Reading the code, there's a comment in the `BaseIntegrationTest` which remarks about a hack where it calls `sleep()` [1]. I think that is where the time system initializes its thread-id for further checking later on.
If it wouldn't do that we would be fine... Unfortunately, the constructor that accepts a time system ignores whatever we pass in, so it seems that we also cannot pass in a different time system that doesn't perform this check to avoid the assert. 

[1] https://github.com/envoyproxy/envoy/blob/b41ba5925a4e93d22a86c6501d63314ccf0d79f3/test/integration/integration.cc#L239

**Comment by oschaaf** [comment]
I think we should land https://github.com/envoyproxy/nighthawk/pull/15 first, and then rebase this.

**Comment by oschaaf** [comment]
Note: just merged master into this, which should clean up the diff now that #15 was merged over there.

**Comment by htuch** [comment]
@oschaaf can you merge master again? Thanks

**Comment by oschaaf** [comment]
@htuch done, merged master

File: ci/run_coverage.sh [hunk 1]
```diff
@@ -32,7 +32,7 @@ if [ "$VALIDATE_COVERAGE" == "true" ]
 then
     COVERAGE_VALUE=$(grep -Po '.*lines[.]*: \K(\d|\.)*' "${COVERAGE_SUMMARY}")
     # TODO(oschaaf): The target is 97.5%, so up this whenever possible in follow ups.
-    COVERAGE_THRESHOLD=96.6
+    COVERAGE_THRESHOLD=96.8
```
Comment (htuch) [inline ci/run_coverage.sh]: Huzzah!

File: source/common/rate_limiter_impl.cc [hunk 2]
```diff
@@ -6,6 +6,39 @@
...
+  if (releasing_) {
+    accumulated_--;
+    releasing_ = accumulated_ > 0;
+    return true;
+  } else {
+    // We are greedy here, which should help with timely releases of bursts when
+    // the system is lagging behind.
+    while (rate_limiter_->tryAcquireOne()) {
+      accumulated_++;
+      if ((accumulated_ % burst_size_) == 0) {
```
Comment (htuch) [inline source/common/rate_limiter_impl.cc]: The use of `%` here together with both ++ on release and acquire is clever; it took me a minute to grok though, could you add a comment or two?

File: source/common/rate_limiter_impl.cc [hunk 3]
```diff
@@ -6,6 +6,39 @@
...
+  }
+  return false;
+}
+
+void BurstingRateLimiter::releaseOne() {
+  if (releasing_) {
+    accumulated_++;
+    ASSERT(accumulated_ <= burst_size_);
+  } else {
+    rate_limiter_->releaseOne();
```
Comment (htuch) [inline source/common/rate_limiter_impl.cc]: Is it ever possible to execute this line?

File: source/common/rate_limiter_impl.h [hunk 4]
```diff
@@ -10,6 +10,28 @@
 
 namespace Nighthawk {
 
+/**
+ * BurstingRatelimiter can be wrapped around another rate limiter. It has two modes:
+ * 1. First it will be accumulating acquisitions by forwarding calls to the wrapped
+ *    rate limiter, until the accumulated acquisitions equals the specified burst size.
+ * 2. Release mode. In this mode, BatchingRateLimiter is in control and will be handling
+ *    acquisition calls (returning true and substracting from the accumulated total until
```
Comment (htuch) [inline source/common/rate_limiter_impl.h]: It cleverly uses increment to subtract! :)

File: test/rate_limiter_test.cc [hunk 5]
```diff
@@ -39,4 +41,78 @@ TEST_F(RateLimiterTest, LinearRateLimiterInvalidArgumentTest) {
...
 }
 
+TEST_F(RateLimiterTest, BurstingRateLimiterTest) {
+  const uint64_t burst_size = 3;
+  std::unique_ptr<MockRateLimiter> mock_rate_limiter = std::make_unique<MockRateLimiter>();
+  MockRateLimiter& unsafe_mock_rate_limiter = *mock_rate_limiter;
+  EXPECT_CALL(unsafe_mock_rate_limiter, tryAcquireOne)
+      .Times(burst_size)
+      .WillRepeatedly(Return(true))
+      .RetiresOnSaturation();
```
Comment (htuch) [inline test/rate_limiter_test.cc]: What is `RetireOnSaturation`? Is this effectively the same as `InSequence`?

File: source/common/rate_limiter_impl.cc (hunk 2 above)
Comment (oschaaf) [reply source/common/rate_limiter_impl.cc]: Done

File: source/common/rate_limiter_impl.cc (hunk 3 above)
Comment (oschaaf) [reply source/common/rate_limiter_impl.cc]: No, good point, this didn't make sense. I tightened this up.

File: test/rate_limiter_test.cc (hunk 5 above)
Comment (oschaaf) [reply test/rate_limiter_test.cc]: Thanks, `InSequence` specifies the intent here more clearly and is more compact, so I switched it over to use that.

**Comment by oschaaf** [comment]
@htuch this is ready for another round

File: source/common/rate_limiter_impl.cc [hunk 6]
```diff
@@ -6,6 +6,50 @@
...
+  }
+
+  previously_releasing_ = absl::nullopt;
+  return false;
+}
+
+void BurstingRateLimiter::releaseOne() {
+  ASSERT(accumulated_ < burst_size_);
+  ASSERT(previously_releasing_ != absl::nullopt && previously_releasing_ == true);
+  // The caller wasn't able to put its earlier successfull acquisition to good use, so we restore
```
Comment (htuch) [inline source/common/rate_limiter_impl.cc]: Oh, this is actually different to what I thought you were doing. I thought you were doing:
* During acquisition, keep increasing accumulated until % burst_size == 0.
* During release, keep increasing accumulated until % burst_size ==0.

If you ++ and --, won't you run the risk of getting non-bursty behavior when released tokens get immediately reacquired and a thread spends a long period without requesting a burst?

File: source/client/factories_impl.cc [hunk 7]
```diff
@@ -41,6 +41,10 @@ SequencerPtr SequencerFactoryImpl::create(Envoy::TimeSource& time_source,
   StatisticFactoryImpl statistic_factory(options_);
   RateLimiterPtr rate_limiter =
       std::make_unique<LinearRateLimiter>(time_source, Frequency(options_.requestsPerSecond()));
+  const uint64_t burst_size = options_.burstSize();
+  if (burst_size) {
```
Comment (htuch) [inline source/client/factories_impl.cc]: Nit: prefer `burst_size > 0`

**Pull request merged**